tsdb/head_append.go

```diff
@@ -358,6 +383,40 @@ func (a *headAppender) AppendExemplar(ref storage.SeriesRef, lset labels.Labels,
 	return storage.SeriesRef(s.ref), nil
 }
 
+// AppendMetadata for headAppender assumes the series ref already exists, and so it doesn't
+// use getOrCreate or make any of the lset sanity checks that Append does.
+func (a *headAppender) AppendMetadata(ref storage.SeriesRef, lset labels.Labels, meta metadata.Metadata) (storage.SeriesRef, error) {
+	s := a.head.series.getByID(chunks.HeadSeriesRef(ref))
+	if s == nil {
+		s = a.head.series.getByHash(lset.Hash(), lset)
+		if s != nil {
+			ref = storage.SeriesRef(s.ref)
+		}
+	}
+	if s == nil {
+		return 0, fmt.Errorf("unknown series when trying to add metadata with HeadSeriesRef: %d and labels: %s", ref, lset)
+	}
+
+	s.RLock()
+	hasNewMetadata := s.meta != meta
+	s.RUnlock()
+
+	if hasNewMetadata {
+		s.Lock()
+		s.meta = meta
```

Comment:
Just realised something here. We should do this in Commit() and only collect the record here. `AppendMetadata` should only collect what all changes we need to do. In Commit() you can simply go over `a.metadata` and apply it. Similar to samples, we can collect `metadataSeries` and use that series in `Commit()`.

Reply:
I've made this change in https://github.com/prometheus/prometheus/pull/10312/commits/e7b116c9da8c9ac6370b55beee88f551384b2d56. It requires an extra conversion from record.RefMetadata to metadata.Metadata, that's why that new helper was added. 

Is that like what you had in mind? ^^